docs:#1010 combine version upgrades into rolling upgrades

micheleRP opened the pull request:
resolves #1010 

This incorporates the Version Upgrades content into the Rolling Upgrades page. (Version Upgrades was removed from our new library structure, so this adds the upgrade info back in.) It also adds some other upgrade info/tips that was missing from our doc. 
- preview 23.1: https://deploy-preview-1246--redpanda-documentation.netlify.app/docs/manage/cluster-maintenance/rolling-upgrade/
- preview 22.3: https://deploy-preview-1246--redpanda-documentation.netlify.app/docs/22.3/manage/cluster-maintenance/rolling-upgrade/

micheleRP commented:
@vuldin, @daisukebe, @dotnwat: Please review this revised page for rolling upgrades. Also, I know we've removed many instances of "vectorized", so please confirm also if it's OK to use:

If you previously installed Redpanda with brew, run:
`brew install vectorizedio/tap/redpanda`

and 
```yaml
...
spec:
  image: "vectorized/redpanda"
  version: "<current_version>"
...
```

JakeSCahill commented:
@micheleRP I think we can take Kubernetes info out of this page. It’s confusing to see maintenance mode tasks when they don’t apply. I’ll be doing a dedicated Kubernetes page as part of https://github.com/redpanda-data/documentation/issues/1170 this sprint.

daisukebe commented on versioned_docs/version-22.3/manage/cluster-maintenance/rolling-upgrade.mdx:
```diff
@@ -7,33 +7,117 @@ title: Rolling Upgrades
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+To benefit from Redpanda's new features and enhancements, upgrade to the latest version. New features are enabled after all nodes (brokers) in the cluster are upgraded.
```
Can we add this? New features are enabled as long as `all` nodes are alive with a given new version. In a case there's a dead node in the cluster, enabling new features won't be triggered, which is not explicit to users.

(forgive me my broken English :wink: )
```
Ensure all nodes are up and running before upgrading the version. Use 'rpk redpanda admin brokers list'.
```

micheleRP replied:
Thanks, @daisukebe! I added this in the Perform a rolling upgrade section intro. Please review: https://deploy-preview-1246--redpanda-documentation.netlify.app/docs/manage/cluster-maintenance/rolling-upgrade/#perform-a-rolling-upgrade

daisukebe commented on versioned_docs/version-22.3/manage/cluster-maintenance/rolling-upgrade.mdx:
```diff
@@ -7,33 +7,117 @@ title: Rolling Upgrades
+
+To benefit from Redpanda's new features and enhancements, upgrade to the latest version. New features are enabled after all nodes (brokers) in the cluster are upgraded.
+
+:::note notes
```
Do we have a definition on; what `one version` mean and what `minor versions` mean by?

micheleRP replied:
Thanks @daisukebe! Please review this updated intro: https://deploy-preview-1246--redpanda-documentation.netlify.app/docs/manage/cluster-maintenance/rolling-upgrade/

daisukebe replied:
Great, thanks!

daisukebe commented on versioned_docs/version-22.3/manage/cluster-maintenance/rolling-upgrade.mdx:
````diff
@@ -7,33 +7,117 @@ title: Rolling Upgrades
+<TabItem value="check-linux" label="Linux" default>
+
+```bash
+curl -s -XGET http://<broker host>:9644/v1/brokers/
````
Assuming all users are running >= v22.1, no more v21.x users, users can identify the broker version using `rpk redpanda admin brokers list`. @mattschumpert, can we assume so?

micheleRP replied:
Great question, @daisukebe. @mattschumpert, and if we have no v21.11 users, can we remove all 21.11 doc; e.g., https://docs.redpanda.com/docs/21.11/home/

micheleRP replied:
Not sure about customers running 21.11. See https://redpandadata.slack.com/archives/C02JDTMG8GN/p1677096032290669

mattschumpert replied:
We should have paying customers off 21.11 docs now, but we should only mark them as EOL at the top of the pages not remove them entirely because its unfriendly to community users on that version who still want to use community edition and/or upgrade eventually.

mattschumpert replied:
However, for current version doc set, we can make the assumption of rpk redpanda admin brokers list for instructional purposes.

daisukebe commented on versioned_docs/version-22.3/manage/cluster-maintenance/rolling-upgrade.mdx:
````diff
@@ -7,33 +7,117 @@ title: Rolling Upgrades
+<TabItem value="check-kubernetes" label="Kubernetes" default>
+
+```bash
+kubectl exec -it -n <cluster_name> <pod-name> -- rpk version
````
Same above. Assuming all users are running >= 22.1, we can use the command below.

```
kubectl exec -it -n <cluster_name> <pod-name> -- rpk redpanda admin brokers list
```

micheleRP replied:
resolved here: https://github.com/redpanda-data/documentation/pull/1297/files#diff-2fac7b96f96c7e0bb400f0a2ee80bc5bce9b3841860bc637189110ef6ed406d4

micheleRP commented:
> #1170 this sprint.

Thanks, @JakeSCahill: I've removed K8s content from this page. Please remind me after your content is in to add a link to it on this page (if you don't).

micheleRP commented:
> @vuldin, @daisukebe, @dotnwat: Please review this revised page for rolling upgrades. Also, I know we've removed many instances of "vectorized", so please confirm also if it's OK to use:
> 
> If you previously installed Redpanda with brew, run: `brew install vectorizedio/tap/redpanda`
> 
> and
> 
> ```yaml
> ...
> spec:
>   image: "vectorized/redpanda"
>   version: "<current_version>"
> ...
> ```

@JakeSCahill The 2nd reference here is moving to your new K8s page. For the 1st reference: I changed this to `brew install redpanda-data/tap/redpanda`. Can you please confirm that that's correct?

JakeSCahill commented:
> > @vuldin, @daisukebe, @dotnwat: Please review this revised page for rolling upgrades. Also, I know we've removed many instances of "vectorized", so please confirm also if it's OK to use:
> > If you previously installed Redpanda with brew, run: `brew install vectorizedio/tap/redpanda`
> > and
> > ```yaml
> > ...
> > spec:
> >   image: "vectorized/redpanda"
> >   version: "<current_version>"
> > ...
> > ```
> 
> @JakeSCahill The 2nd reference here is moving to your new K8s page. For the 1st reference: I changed this to `brew install redpanda-data/tap/redpanda`. Can you please confirm that that's correct?

`brew install redpanda-data/tap/redpanda` works

micheleRP commented on versioned_docs/version-22.3/manage/cluster-maintenance/rolling-upgrade.mdx:
```diff
@@ -68,7 +165,94 @@ Repeat this process for each node to upgrade.
+</TabItem>
+
+</Tabs>
+
+### Disable maintenance mode
```
@RafalKorepta Issue #1061 says to add a description here of what to metrics to monitor at this step. Our doc hasn't yet documented vectorized_cluster_partition_leader_id or vectorized_cluster_partition_under_replicated_replicas, and I'm not sure which I/O metrics to add here. What general recommendations should we include (and then we can add links to specific metrics/APIs when we have doc for them)?

RafalKorepta replied:
Thanks to @jcsp ([PR-9011](https://github.com/redpanda-data/redpanda/pull/9011)) the cluster health will have more useful information starting from v23.1.x.

Before this version I would suggest to watch for:

- any none zero value for metrics `vectorized_cluster_partition_under_replicated_replicas`. That would indicate replication can not catch up for various reasons and the upgrade should be paused.
- if vectorized_cluster_partition_leader_id for any partition has -1 value. That would indicate we lost partition leader and we can not produce to such partition
- IO metrics:
  - vectorized_storage_log_written_bytes - The rate for each broker should go to circa value before restart
  - vectorized_storage_log_read_bytes - The rate for each broker should go to circa value before restart
  - vectorized_raft_leadership_changes - after restart eventually it should go to 0
  - vectorized_raft_leader_for - the sum for each broker should be equal. Eventually
  - vectorized_kafka_latency_produce_latency_us_bucket - p99 histogram should go to circa value before restart
  - vectorized_kafka_latency_fetch_latency_us_bucket - p99 histogram should go to circa value before restart
  - vectorized_kafka_rpc_dispatch_handler_latency_bucket - p99 histogram should go to circa value before restart


If @jcsp or @esteban have more details to share please feel free to correct me :)

micheleRP replied:
Thanks @RafalKorepta! Please check this new step: https://deploy-preview-1246--redpanda-documentation.netlify.app/docs/manage/cluster-maintenance/rolling-upgrade/#check-metrics. And a couple questions:
- I don't understand "go to circa value". Could I change this "return to the original value"? If so, should users check these metrics before running the upgrade too? 
- This is going out with the 23.1 docs. Can you please confirm that we should include this section (i.e., that checking `rpk cluster health` is not enough)?

RafalKorepta replied:
> Could I change this "return to the original value"? If so, should users check these metrics before running the upgrade too?

The `original value` for me doesn't bring the context that value should come back to the value before an upgrade. Never the less my English is not that good :) 

> This is going out with the 23.1 docs. Can you please confirm that we should include this section (i.e., that checking rpk cluster health is not enough)?

The `rpk cluster health` is not enough as I think metrics should be watched too. I will let @jcsp confirm this.

RafalKorepta commented on versioned_docs/version-22.3/manage/cluster-maintenance/rolling-upgrade.mdx:
````diff
@@ -7,33 +7,112 @@ title: Rolling Upgrades
+:::
+
+```bash
+docker exec -it <container_name> rpk version 
````
I would be nit pick this command. The container_name variable should consist of 

1. container name
2. container tag
The container tag would determine what version of rpk is used. Not necessary the Redpanda binary. Fortunately the release process is construct in that way we bundle rpk and Redpanda into the same container tag with the same version. 

If you would like to list the available docker hub tags the following command:
```bash
curl -s 'https://hub.docker.com/v2/repositories/vectorized/redpanda/tags/?ordering=last_updated&page=1&page_size=50' | jq -r '.results[].name'
```  
would return:
```
v22.3.13
latest
v22.3.13-arm64
v22.3.13-amd64
v22.2.10
v22.2.10-arm64
v22.2.10-amd64
v22.3.12
v22.3.12-arm64
v22.3.12-amd64
v22.3.11
v22.3.11-arm64
v22.3.11-amd64
v22.3.10
v22.3.10-arm64
v22.3.10-amd64
v22.1.11
v22.1.11-arm64
v22.1.11-amd64
v22.2.9
v22.2.9-arm64
v22.2.9-amd64
v22.3.9
v22.3.9-arm64
v22.3.9-amd64
v22.3.8
v22.3.8-arm64
v22.3.8-amd64
v22.3.6
v22.3.6-arm64
v22.3.6-amd64
v22.3.5
v22.3.5-arm64
v22.3.5-amd64
v22.2.8
v22.2.8-arm64
v22.2.8-amd64
v22.3.4
v22.3.4-arm64
v22.3.4-amd64
v22.1.10
v22.1.10-arm64
v22.1.10-amd64
v21.11.22
latest-arm64
v21.11.22-arm64
latest-amd64
v21.11.22-amd64
v21.11.22-beta2
v21.11.22-beta2-arm64
```

You can always adjust the number of result using `page_size`.

If our user would like to see v22 available version, they could use the following command:
```bash
curl -s 'https://hub.docker.com/v2/repositories/vectorized/redpanda/tags/?name=v22&ordering=last_updated&page=1&page_size=25' | jq -r '.results[].name'
```

The name parameter `v22` could be changed to what ever release train we have e.g. `v23` or `v22.3.`

micheleRP replied:
Thanks @RafalKorepta! Please review https://deploy-preview-1246--redpanda-documentation.netlify.app/docs/manage/cluster-maintenance/rolling-upgrade/#find-your-current-version

RafalKorepta replied:
Great!

RafalKorepta commented on versioned_docs/version-22.3/manage/cluster-maintenance/rolling-upgrade.mdx:
````diff
@@ -7,33 +7,112 @@ title: Rolling Upgrades
+To ensure that all nodes are active before upgrading, run:
+
+```bash 
+rpk redpanda admin brokers list
````
Maybe we should use 
```
rpk cluster health
```

This command would return:
```
CLUSTER HEALTH OVERVIEW
=======================
Healthy:               true
Controller ID:               0
All nodes:                   [1 2 0]
Nodes down:                  []
Leaderless partitions:       []
Under-replicated partitions: []
```

This command would be available starting from v22.1.1

RafalKorepta replied:
Oh I see that it's addressed in the next headline `Enable maintenance mode`

RafalKorepta commented on versioned_docs/version-22.3/manage/cluster-maintenance/rolling-upgrade.mdx:
```diff
@@ -7,33 +7,112 @@ title: Rolling Upgrades
+The logical cluster version is incremented after the last active node is upgraded. If problems occur, the upgrade is not committed.
+
+:::note
+After all nodes are upgraded, Redpanda starts to migrate consumer group topics to __consumer_offsets. This takes some time, depending on the data size. Until it finishes, all consumer group related operations (consume, offset commit, coordinator election) are blocked.
```
This apply only to upgrades from v21.11 to v22.1 according to the @daisukebe issue. REF https://github.com/redpanda-data/documentation/issues/415

> As soon as all nodes are up and running with 22.1, it starts to migrate the consumer group topics to __consumer_offsets, which is a new feature introduced in 22.1 for Kafka compatibility. It will take some time depending on the data size, and until it finishes all consumer group related operations (consume, offset commit, coordinator election, etc) will be blocked.


I'm not sure if this should be precisely mention `from v21.11.x to v22.1.x`

daisukebe replied:
> I'm not sure if this should be precisely mention from v21.11.x to v22.1.x

Agreed.

micheleRP replied:
Thanks @RafalKorepta and @daisukebe. Is this note still relevant to current customers (e.g., customers upgrading to 23.1), or should we remove the note altogether? 

After all nodes are upgraded, Redpanda starts to migrate consumer group topics to __consumer_offsets. This takes some time, depending on the data size. Until it finishes, all consumer group related operations (consume, offset commit, coordinator election) are blocked.

daisukebe replied:
@micheleRP , we may want to leave it as is to be friendly for users still running 21.11 or lower. However, the logic only applies to the one time upgrade from 21.11 to 22.1. We should note it. When upgrading from 22.1 to 22.2 and further, this isn't the case any more.

RafalKorepta commented on versioned_docs/version-22.3/manage/cluster-maintenance/rolling-upgrade.mdx:
````diff
@@ -7,33 +7,112 @@ title: Rolling Upgrades
+    </summary>
+
+    ```bash
+    Healthy: true
````
NIT:

Could we show whole output but highlight only the `Healthy: true`?

micheleRP replied:
@RafalKorepta I added more expected output many of the steps. Let me know if it's too much!

RafalKorepta commented on versioned_docs/version-22.3/manage/cluster-maintenance/rolling-upgrade.mdx:
````diff
@@ -48,16 +127,34 @@ Repeat this process for each node to upgrade.
+    </summary>
+
+    ```bash
+    FINISHED: true
````
ditto

whole output with highlight of `FINISHED: true`

RafalKorepta commented on versioned_docs/version-22.3/manage/cluster-maintenance/rolling-upgrade.mdx:
````diff
@@ -48,16 +127,34 @@ Repeat this process for each node to upgrade.
+    </summary>
+
+    ```bash
+    Healthy: true
````
ditto

RafalKorepta commented on versioned_docs/version-22.3/manage/cluster-maintenance/rolling-upgrade.mdx:
````diff
@@ -68,7 +165,94 @@ Repeat this process for each node to upgrade.
+If you previously installed Redpanda with brew, run:
+
+```bash
+brew install redpanda-data/tap/redpanda
````
I'm not sure if `install` is the right command for an upgrade. I'm personally used to 
```
brew upgrade ...
```

RafalKorepta commented on versioned_docs/version-22.3/manage/cluster-maintenance/rolling-upgrade.mdx:
````diff
@@ -68,7 +165,94 @@ Repeat this process for each node to upgrade.
+docker restart <container_name>
+```
+
+For more information, see [Redpanda Quickstart](../../../get-started/quick-start/quick-start-docker).
````
I'm not sure if we should advise user to use docker compose or even `rpk container start` for testing:
```
rpk container start --help
Start a local container cluster

Usage:
  rpk container start [flags]

Flags:
  -h, --help           Help for start
  -n, --nodes uint     The number of nodes to start (default 1)
      --retries uint   The amount of times to check for the cluster before considering it unstable and exiting. (default 10)

Global Flags:
  -v, --verbose   Enable verbose logging (default: false)
```

or 
https://github.com/redpanda-data/documentation/blob/e289b8f91e617be77c2313ccc734c2742f4a7966/docs/get-started/quick-start.mdx?plain=1#L121-L191
or
https://github.com/redpanda-data/documentation/blob/e289b8f91e617be77c2313ccc734c2742f4a7966/docs/get-started/quick-start.mdx?plain=1#L225-L355

micheleRP replied:
for reference: https://redpandadata.slack.com/archives/C04LYNJQUGN/p1677794110689209

RafalKorepta commented on versioned_docs/version-22.3/manage/cluster-maintenance/rolling-upgrade.mdx:
````diff
@@ -84,10 +268,26 @@ After you've successfully upgraded the node:
     ```
 
-### 3. Restart node
+### Restart node
````
I'm a bit confused. The restart should be before disabling maintenance mode.

micheleRP replied:
Thanks @RafalKorepta. Please confirm if this now looks correct: https://deploy-preview-1246--redpanda-documentation.netlify.app/docs/manage/cluster-maintenance/rolling-upgrade/#perform-a-rolling-upgrade

RafalKorepta replied:
LGTM

RafalKorepta reviewed (commented):
I left few comments

JakeSCahill reviewed (approved):
Other than the existing open questions in the comments, LGTM. @micheleRP we should add the impact of restarts partial to this doc: https://github.com/redpanda-data/documentation/tree/dev/docs/manage/shared/rolling-upgrades

micheleRP commented:
> Other than the existing open questions in the comments, LGTM.
> 
> @micheleRP we should add the impact of restarts partial to this doc: https://github.com/redpanda-data/documentation/tree/dev/docs/manage/shared/rolling-upgrades
> 
> Also, do we need to forwardport these changes to 23.1?
> 
> The build is failing ATM because:
> 
> ```
> 7:30:50 PM: - On source page path = /docs/22.3/manage/cluster-maintenance/rolling-upgrade/:
> 7:30:50 PM:    -> linking to ../../../get-started/quick-start/quick-start-docker/ (resolved as: /docs/22.3/get-started/quick-start/quick-start-docker/)
> ```

Thanks, @JakeSCahill! I added the shared file to 22.3 and also made some more edits to be consistent with your K8s version.

micheleRP commented:
@mattschumpert, @RafalKorepta, @daisukebe, @vuldin, @dotnwat: This PR now also incorporates the updates in https://github.com/redpanda-data/documentation/pull/1297. Please give this a final review!

RafalKorepta commented on versioned_docs/version-22.3/manage/shared/rolling-upgrades/_restart-impact.mdx:
```diff
@@ -0,0 +1,18 @@
+
+### CPU spikes upon broker restart
+
+When a restarted broker becomes available again, you may see your nodes' CPU usage increase temporarily. This temporary increase in CPU usage is due to the cluster rebalancing the partition replicas.
```
Recent incident brought good point that production setup for Redpanda will always use around 80-95% CPU as [Seastar](https://seastar.io/) is always pooling. It might be misleading for users. They should look for `vectorized_reactor_utilization` metrics. @daisukebe could know if there is more dimensions/metrics to watch for.

micheleRP replied:
and please see John's comment below that we shouldn't encourage users to use the internal metrics endpoint. Should they check [redpanda_uptime_seconds_total ](https://docs.redpanda.com/docs/reference/public-metrics-reference/#redpanda_uptime_seconds_total) instead?

daisukebe replied:
@jcsp , should we encourage users to monitor `redpanda_cpu_busy_seconds_total` instead of `vectorized_reactor_utilization`? If that's the case, do we recommend them to take the derivative as you mentioned in https://redpandadata.slack.com/archives/C01ND4SVB6Z/p1669395833531899?thread_ts=1669392271.380349&cid=C01ND4SVB6Z?

jcsp replied:
`redpanda_cpu_busy_seconds_total` is indeed the place to monitor CPU utilization.

micheleRP replied:
Thanks @jcsp and @RafalKorepta. Please see the updated list of metrics: https://deploy-preview-1246--redpanda-documentation.netlify.app/docs/manage/cluster-maintenance/rolling-upgrade/#check-metrics, and please confirm what they should look for with `redpanda_cpu_busy_seconds_total`. Should it be similar to what we have here:   https://docs.redpanda.com/docs/manage/monitoring/#infrastructure-resources

RafalKorepta replied:
LGTM

jcsp commented:
Everywhere this page mentions a `vectorized_` metric, it should be mentioning a `redpanda_` metric (or omit it if there is no public metric) -- we do not want to encourage users to use the internal metrics endpoint.

micheleRP commented:
> > If @jcsp or @esteban have more details to share please feel free to correct me :)
> 
> Everywhere this page mentions a `vectorized_` metric, it should be mentioning a `redpanda_` metric (or omit it if there is no public metric) -- we do not want to encourage users to use the internal metrics endpoint.

Thanks @jcsp. Would you please confirm if this looks correct: https://deploy-preview-1246--redpanda-documentation.netlify.app/docs/manage/cluster-maintenance/rolling-upgrade/#check-metrics

hcoyote commented on docs/manage/cluster-maintenance/rolling-upgrade.mdx:
```diff
@@ -11,95 +11,425 @@ contextLinks:
+
+:::note notes
+- Redpanda only supports upgrading one sequential feature release at a time. For example, you can upgrade from the 22.2 feature release to 22.3. You cannot skip feature releases. 
+- Redpanda only supports downgrading between sequential patch releases. For example, you can downgrade from the 22.2.2 patch release to 22.2.1, but you cannot downgrade to 22.1.7.
```
Thes two statements imply that if we upgrade from 22.1.<latest> to 22.2.latest to move between feature releases, that we cannot do a rollback in the event we need to do a downgrade. Meaning, we can only roll-forward. That doesn't seem right and will present issues to customers.

micheleRP replied:
@mattschumpert can you please confirm if this note is correct. And if so, if Travis's comment is also correct?

mattschumpert replied:
The above is corrrect

mattschumpert replied:
@hcoyote we don't support major version downgrade once the last node in the cluster has been upgraded (and the logical version in our metadata is bumped, new features get activated, etc.). We never have.  There's some interest to do so but not particularly strong.

hcoyote replied:
We need to call that out then, if the only safe way to move backwards is if we haven't finalized the version upgrade on all brokers. and I'd elevate it to an important or warning block if we can do that.

hcoyote commented on versioned_docs/version-22.3/manage/shared/rolling-upgrades/_restart-impact.mdx:
```diff
@@ -0,0 +1,18 @@
+When brokers restart, clients may experience higher latency, nodes may experience CPU spikes when the broker becomes available again, and you may receive alerts about under-replicated partitions.
```
Related to comment earlier about single-replica topics. The impact of restarts should also call out that topics using replication factor = 1 would be unavailable. I would classify this as somewhat different than under-replicated partitions. 

we make a comment in rolling-upgrade.mdx about repl=1 and probably need to address that here as well. While it's not something we recommend people configuring for, if they do it, they'll end up with a data access outage.

micheleRP replied:
@hcoyote do you think we should just add to the end of that paragraph: "Topics that weren't using replication (that is, topics that had `replication.factor=1`) will be unavailable." @JakeSCahill since you worked on this!

hcoyote replied:
Yeah, something like that would be good.

hcoyote commented on versioned_docs/version-22.3/manage/cluster-maintenance/rolling-upgrade.mdx:
````diff
@@ -4,90 +4,421 @@ title: Rolling Upgrades
+docker exec -it <container_name><container_tag> rpk version 
+```
+
+Remember to replace the variables `<container_name>` and `container_tag`. The container tag determines which version of `rpk` to use. The release process bundles `rpk` and `Redpanda` into the same container tag with the same version. 
````
this should probably be `<container_tag>` for formatting consistency with the command line above.

hcoyote commented on docs/manage/cluster-maintenance/rolling-upgrade.mdx:
````diff
@@ -11,95 +11,425 @@ contextLinks:
+docker exec -it <container_name><container_tag> rpk version 
+```
+
+Remember to replace the variables `<container_name>` and `container_tag`. The container tag determines which version of `rpk` to use. The release process bundles `rpk` and `Redpanda` into the same container tag with the same version. 
````
this should probably be <container_tag> for formatting consistency with the command line above.

hcoyote commented on docs/manage/cluster-maintenance/rolling-upgrade.mdx:
````diff
@@ -11,95 +11,425 @@ contextLinks:
 
 :::note
-If you are using the Redpanda Kubernetes Operator to trigger an upgrade, then node maintenance mode is automatically incorporated into the upgrade process.
+If your current version is more than one feature release behind the latest Redpanda version, you must first upgrade to an intermediate version. To list all available versions:
+
+```bash
+curl -s 'https://hub.docker.com/v2/repositories/redpandadata/redpanda/tags/?ordering=last_updated&page=1&page_size=50' | jq -r '.results[].name'
+```
+
+<details>
+<summary>
+Example output
+</summary>
+
+```
+v22.3.13
+latest
+v22.3.13-arm64
+v22.3.13-amd64
+v22.2.10
+v22.2.10-arm64
+v22.2.10-amd64
+v22.3.12
+v22.3.11
+v22.3.10
+...
+```
+
+</details>
+
````
This only makes sense in the context of a docker setup. If you're looking at the Linux or MacOS tabs, the available version format strings are going to be different because it's going to depend on the OS package management tool.

micheleRP replied:
I've removed this example output. The doc team still needs to standardize on how much example output we include and the format for it. This example could be more confusing than helpful.

hcoyote commented on docs/manage/cluster-maintenance/rolling-upgrade.mdx:
```diff
@@ -11,95 +11,425 @@ contextLinks:
-1. [Enable maintenance mode](#1-enable-maintenance-mode).
-2. [Disable maintenance mode](#2-disable-maintenance-mode).
-3. [Restart the node](#3-restart-node).
+To ensure that all brokers are active before upgrading, run:
```
probably want to note that they need to look at the Membership-Status column, just in case none of those entries say "active". A pointer to what to do if they're not active might be good as well.

micheleRP replied:
@hcoyote what should they do if one is not active?

hcoyote commented on docs/manage/cluster-maintenance/rolling-upgrade.mdx:
```diff
@@ -11,95 +11,425 @@ contextLinks:
+
+</details>
+
+The logical cluster version is incremented after the last active broker is upgraded. If problems occur, the upgrade is not committed.
```
Is this talking about the broker-version in the previous output or the idea of the cluster config version?

daisukebe replied:
It is talking about a Redpanda internal logical version, which tracks new features. It's not either a broker-version or a cluster config version at all. I guess it's not a good idea to mention about `logical cluster version` unless we explicitly explains it somewhere. It just confuses users. Rather can we just say something like below? @jcsp 

```
A new product feature, that's available in a given feature release, will be fully activated after the last active broker is upgraded. If problems occur, the upgrade is not committed.
```

micheleRP replied:
I updated this from: 
The logical cluster version is incremented after the last active broker is upgraded.
to 
New features in a version are enabled after all brokers in the cluster are upgraded.

hcoyote commented on docs/manage/cluster-maintenance/rolling-upgrade.mdx:
```diff
@@ -11,95 +11,425 @@ contextLinks:
+The logical cluster version is incremented after the last active broker is upgraded. If problems occur, the upgrade is not committed.
+
+:::note
+When upgrading from version 21.11 to 22.1, after all brokers are upgraded, Redpanda starts to migrate consumer group topics to __consumer_offsets. This takes some time, depending on the data size. Until it finishes, all consumer group related operations (consume, offset commit, coordinator election) are blocked.
```
How can they track this or know when it's completed?

micheleRP replied:
@daisukebe for this use case, do you know how to track when consumer groups are done migrating?

daisukebe replied:
Good question. Once you see `consumer offset feature enabled` in all nodes, we can assume it's done.

hcoyote commented on docs/manage/cluster-maintenance/rolling-upgrade.mdx:
```diff
@@ -11,95 +11,425 @@ contextLinks:
+
+### Restart broker
+
+After you're out of maintenance mode, restart the broker's Redpanda service.
```
Wouldn't we restart this before we exit maintenance mode?

hcoyote commented on docs/manage/cluster-maintenance/rolling-upgrade.mdx:
```diff
@@ -11,95 +11,425 @@ contextLinks:
+After you're out of maintenance mode, restart the broker's Redpanda service.
+
+:::note
+You can stop the upgrade process and roll back to the original version as long as you have not upgraded every broker and restarted the cluster.
```
This seems like an important caveat that we should list higher up in the doc. Maybe in the initial intro section or under the pre-reqs as part of the understanding people need to work through.

hcoyote reviewed (commented):
I ran through the generic rolling-upgrade.mdx and skimmed over the 22.3 version ... probably everything I commented on in the generic applies to the versioned doc as well.

micheleRP commented on docs/manage/cluster-maintenance/rolling-upgrade.mdx:
```diff
@@ -11,95 +11,425 @@ contextLinks:
-One of the primary uses for maintenance mode is to perform a rolling upgrade on each node in the cluster. This process involves putting a node into maintenance mode, upgrading the node, taking the node out of maintenance mode, and then repeating the process on the next node in the cluster. Placing nodes into maintenance mode ensures a smooth upgrade of your cluster while reducing the risk of inte ...
+To benefit from Redpanda's new features and enhancements, upgrade to the latest version. New features are enabled after all brokers (nodes) in the cluster are upgraded. 
+
+Rolling upgrades require all brokers to be placed into [maintenance mode](../../node-management) and restarted separately, one after the other. Placing brokers into maintenance mode ensures a smooth upgrade of your cluster while reducing the risk of interruption or degradation in service. When a broker is placed into maintenance mode, it reassigns its partition leadership to other brokers for all ...
```
@hcoyote I believe all the caveats we know about are now in this doc! Should we recommend that for topics with `replication.factor=1`, they increase their replication factor before upgrade? cc @JakeSCahill

hcoyote replied:
+1

hcoyote replied:
I'm not sure what the right answer is for this. 

It's going to depend on availability of free space on the cluster. Cluster operators may have a good reason not to.

You could say something like:

> If you have topics with replication.factor=1, Redpanda recommends temporarily increasing that if you have sufficient free disk space. This can help limit outages for this topic during the rolling upgrade. This operation would need to be done ahead of the rolling restart/upgrade to ensure sufficient time for the data to replicate to other brokers.

micheleRP replied:
Thanks, @hcoyote! I updated the intro to include this.

daisukebe commented on docs/manage/cluster-maintenance/rolling-upgrade.mdx:
````diff
@@ -11,95 +11,427 @@ contextLinks:
     ```
 
-    The `--wait` option ensures that the cluster is healthy before putting the node into maintenance mode.
+    The `--wait` option ensures that the cluster is healthy before putting the broker into maintenance mode. This example puts broker 0 into maintenance mode.
````
> The `--wait` option ensures that the cluster is healthy before putting the broker into maintenance mode.

This is not accurate. Rather we should say something like below.

> The `--wait` option leaves the command as is and waits until a given node, 0 in this example, finishes draining all partitions it originally served. Once the partition draining completes, the command finished.

daisukebe commented on docs/manage/cluster-maintenance/rolling-upgrade.mdx:
````diff
@@ -11,95 +11,427 @@ contextLinks:
+```
 
-### 1. Enable maintenance mode
+All brokers should list `active` in the `MEMBERSHIP-STATUS` column:
````
> All brokers should list `active` in the `MEMBERSHIP-STATUS` column:

and should list `true` in the `IS-ALIVE` column too

daisukebe commented on docs/manage/cluster-maintenance/rolling-upgrade.mdx:
````diff
@@ -11,95 +11,427 @@ contextLinks:
+</summary>
+
+```bash
+v22.3.11 (rev 9eefb90)
````
The example output of `rpk redpanda admin brokers list` in Linux is below, which differs from the commands in MacOS and Docker. 

```
NODE-ID  NUM-CORES  MEMBERSHIP-STATUS  IS-ALIVE  BROKER-VERSION
0        1          active             true      v22.3.11 - 9eefb907c43bf1cfeb0783808c224385c857c0d4-dirty
1        1          active             true      v22.3.11 - 9eefb907c43bf1cfeb0783808c224385c857c0d4-dirty
2        1          active             true      v22.3.11 - 9eefb907c43bf1cfeb0783808c224385c857c0d4-dirty
```

daisukebe replied:
@micheleRP, are you on it? This is still showing a wrong example output for Linux per https://deploy-preview-1246--redpanda-documentation.netlify.app/docs/manage/cluster-maintenance/rolling-upgrade/.

micheleRP replied:
Hi @daisukebe! This issue of showing different output for docker/macOS/linux is the only remaining item on this PR. @hcoyote has a similar comment above. Can either of you can provide the output I should add for macOS & linux? If not, and if the sample docker output is confusing, I'd like to just remove this sample output... since it could be hard to maintain the file going forward for all 3 (and I also don't want to set the precedence that we're going to provide sample output for multiple OS in other files).

daisukebe replied:
@micheleRP, thanks for tracking it!
Well, I'm with you. Given we don't support either MacOS or Docker based installations in a production use, it doesn't basically make sense to guide the users to do a rolling upgrade for those deployments. IIRC, John or Travic mentioned the same earlier in this PR.

@mattschumpert, should we leave the MacOS and Docker options as is in this doc?

micheleRP replied:
@daisukebe I'm going to leave these in for now, since users could possibly upgrade a dev cluster. The doc team still needs to standardize on how much example output we want to provide and the format for it. For now, I'm going to merge this PR. Please feel free to add a new doc issue if you want more updates to this page, or for issues about how we provide example output.

jcsp commented on docs/manage/cluster-maintenance/rolling-upgrade.mdx:
```diff
@@ -11,95 +11,427 @@ contextLinks:
+| [redpanda_kafka_under_replicated_replicas](../../../reference/public-metrics-reference/#redpanda_kafka_under_replicated_replicas) | If this shows any non-zero value, then replication cannot catch up, and the upgrade should be paused. |
+| [redpanda_cluster_unavailable_partitions](../../../reference/public-metrics-reference/#redpanda_cluster_unavailable_partitions)  | If any partition shows a value of `-1`, then the partition leader is lost, and Redpanda cannot produce to it.   |  
+| [redpanda_kafka_request_bytes_total](../../../reference/public-metrics-reference/#redpanda_kafka_request_bytes_total)| Before restart, the produce and consume rate for each broker should go to circa value. |
+| [redpanda_raft_leadership_changes](../../../reference/public-metrics-reference/#redpanda_raft_leadership_changes)  | After restart, this value should eventually go to 0. |      
```
This is a counter: it won't go to zero, but it will stop changing.

I don't know that we necessarily need to ask people to do this though.  It is okay to proceed with an upgrade while some leader transfers are happening.

jcsp commented on versioned_docs/version-22.3/manage/cluster-maintenance/rolling-upgrade.mdx:
```diff
@@ -4,90 +4,423 @@ title: Rolling Upgrades
+For all available flags, see the [`rpk redpanda admin brokers list` command reference](../../../reference/rpk/rpk-redpanda/rpk-redpanda-admin-brokers-list).
+
+</TabItem>
+<TabItem value="check-macos" label="MacOS" default>
```
It's a bit odd to have MacOS instructions here -- afaik we do not support redpanda on MacOS, and the `brew` package is just rpk.  So there's no concept of a rolling upgrade there.

micheleRP replied:
See https://github.com/redpanda-data/console/releases, and we cover MacOS in other sections of the doc. But I'm not on Mac, so I can't provide sample output. It'd be easier for maintaining this file if we don't include MacOS here. What do you think?

mattschumpert commented:
@micheleRP 

I definitely like that this content has been combined, cleaned up, etc. Rolling upgrade is the base case of how customers want to upgrade so it certainly should be on the same page (our main upgrade page).

That said I still find it odd that the page isn't entitled 'Upgrading a Cluster' or better yet, just 'Upgrade/Upgrades'.  If I'm looking to 'Upgrade' thats what my eyes are scanning for.   'Rolling Upgrade' is a detail of how we suggest customers do an upgrade on a production cluster. But It's also perfectly possible to do a hard upgrade (non-rolling) of, for example, a dev cluster, where you really don't care if its rolling because nothing is running on it and you don't want to spend the time doing a rolling upgrade (especially if not on k8s where it kind of happens for you).

On this page, I'd expect to see the an overview explaining how upgrades work overall (e.g. upgrade isn't finalized until the last node is upgraded), the kernel of how to do upgrade a single node, and then explanation/example of rolling upgrade.  Instead this jumps straight into rolling upgrade, which feels slightly awkward.

mattschumpert reviewed (approved):
Minor issue is that the 'restart brokers' step doesn't show the actual command to do so.

kbatuigas commented on modules/ROOT/attachments/cloud-controlplane-api.yaml:
```diff
@@ -331,6 +325,12 @@ components:
           example: development-cluster
           type: string
+        nat_gateways:
+          description: Nat gateway information of the cluster.
```
```suggestion
          description: NAT gateway information of the cluster.
```
